### ISSUE TITLE
Output-to-file cmdlets should not support wildcards with -Path

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#6714]
Note: This is technically a breaking change; if it is deemed too risky, keep it on file for PowerShell vNext.

Currently, output-to-file cmdlets such as `Out-File` and `Set-Content` do support wildcard expression passed to their `-Path` parameter.

This behavior is problematic in two respects:

* Overwriting existing files should be a very deliberate act; the use of wildcards to target _whatever happens to match_  contradicts that.

* The utility of targeting an output file implicitly by wildcard is very limited:

   * If no existing file matches the wildcard expressions, an error occurs.
   * If more than one file matches the wildcard expression, the behavior is _inconsistent_:
       * With _some_ cmdlets, an error occurs (e.g., `Out-File`, `Export-FormatData`
           * In other words: only if the wildcard expression happens to match _one_ existing file does the operation succeed - at which point the clearer and safer choice is to target that file with a _literal_ path.
      * Other cmdlets (e.g.`Set-Content`) output to _all_ matching files, which is also problematic.

Ergo: output-to-file cmdlets should treat `-Path` as an alias of `-LiteralPath` (similar to how `-Path` in `New-Item` is effectively `-LiteralPath`).

Environment data
----------------

Written as of:

```powershell
PowerShell Core v6.0.2
```

**Comment by mklement0** [comment]
While it makes sense to distinguish between interactive and script use with respect to _syntax_, I think extending that distinction to _features_ is problematic.

Yes, you can view doing something like

```powershell
'hi' | Set-Content *.txt
```

to overwrite _all_ `*.txt` files in the current folder as _convenient_, but I think of it as _risky_, and the use case strikes me as too exotic to outweigh the risks.

At the very least, all output-to-file cmdlets should act _consistently_ in such a scenario, in favor of the only-output-only-to-ONE-file behavior (as `Out-File` currently does) ; even documenting the differences won't do, in my opinion,  because it's just too hard to remember which cmdlet behaves how.

**Comment by ThreeFive-O** [comment]
I once started a similar discussion #3174 which was committee reviewed therefore.

While my personal preference would be to treat them as literal paths, I think the current behavior is valid for scripting.

Having wildcard support for input and output even makes it more consistent.

BUT, all cmdlets should behave the same! Apart from file path wildcards, literal paths should be available too. I would go as far to say -Path is ALWAYS with wildcard support and -LiteralPath is a mandatory override.

I have lots of files with square brackets. The way PS handles paths with wildcards always gets me. But that was the initial design of PS and people are using it that way. Maybe something to re-consider in PS7 😀

**Comment by mklement0** [comment]
@BrucePay: Good point about the wildcard in a _directory_ component; that's perhaps a better use case.

I'll close this issue to discuss the problematic `Set-Content` behavior separately.

@ThreeFive-O:

> BUT, all cmdlets should behave the same!

Agreed.

> -Path is ALWAYS with wildcard support and -LiteralPath is a mandatory override.

This distinction is worth observing in general, but there are exceptions, such as with `New-Item`, where wildcards aren't supported (and make no sense); however, `-LiteralPath` should be introduced as an _alias_ parameter name there.

> The way PS handles paths with wildcards always gets me. But that was the initial design of PS and people are using it that way. Maybe something to re-consider in PS7

@ThreeFive-O, @BrucePay:

Yes, the great `-LiteralPath` / `-Path` schism is unfortunate and a perennial pain point.

>  so you have to double-quote all wildcards. 

@BrucePay, I assume you meant `` ` ``-escaping individual wildcard characters inside a _single_-quoted string (you can also use a double-quoted string, but then you must _double_ the `` ` `` chars;  the alternative is to use `[System.Management.Automation.WildcardPattern]::Escape()`):

```powershell
New-Item -Type File 'tmp[a].txt'  # create file whose name *looks* like a wildcard pattern

# BROKEN, because even though the path matches literally, it is interpreted as a wildcard pattern (by default, `-Path` implied).
# What matters is what Get-Item ends up seeing *after* string expansion / quote removal.
Get-Item tmp[a].txt  # same as: Get-Item -Path tmp[a].txt
Get-Item 'tmp[a].txt'  # same
Get-Item "tmp[a].txt"  # same

# OK: All of the following target file 'tmp[a].txt' by its exact, literal name:
Get-Item -Path        'tmp`[a`].txt' # Note the '...' and the individually `-escaped wildcard chars.
Get-Item -Path        "tmp``[a``].txt" # Note the '"..." and the *doubly* individually `-escaped wildcard chars.
Get-Item -LiteralPath  tmp[a].txt    # No quoting necessary
Get-Item -Path ([System.Management.Automation.WildcardPattern]::Escape('tmp[a].txt')) # API use
```

**Comment by mklement0** [comment]
Closed in favor of #6729.
